metrics-recorder-prometheus/src/lib.rs

```diff
@@ -29,80 +32,121 @@ impl PrometheusRecorder {
+        self.output.push_str(name.as_str());
         self.output.push_str(" counter\n");
-        self.output.push_str(label.as_str());
+        self.output.push_str(full_name.as_str());
         self.output.push_str(" ");
         self.output.push_str(value.to_string().as_str());
         self.output.push_str("\n");
     }
 
-    fn record_gauge<K: Into<Key>>(&mut self, key: K, value: i64) {
-        let label = key.into().as_ref().replace('.', "_");
+    fn record_gauge(&mut self, key: Key, value: i64) {
+        let (name, labels) = key_to_parts(key);
+        let full_name = render_labeled_name(&name, &labels);
         self.output.push_str("\n# TYPE ");
-        self.output.push_str(label.as_str());
+        self.output.push_str(name.as_str());
         self.output.push_str(" gauge\n");
-        self.output.push_str(label.as_str());
+        self.output.push_str(full_name.as_str());
         self.output.push_str(" ");
         self.output.push_str(value.to_string().as_str());
         self.output.push_str("\n");
     }
 
-    fn record_histogram<K: Into<Key>>(&mut self, key: K, values: &[u64]) {
-        let mut sum = 0;
-        let mut h = Histogram::<u64>::new(3).expect("failed to create histogram");
+    fn record_histogram(&mut self, key: Key, values: &[u64]) {
+        let entry = self.histos.entry(key).or_insert_with(|| {
+            let h = Histogram::<u64>::new(3).expect("failed to create histogram");
+            (0, h)
+        });
+
+        let (sum, h) = entry;
         for value in values {
             h.record(*value).expect("failed to record histogram value");
-            sum += *value;
-        }
-
-        let label = key.into().as_ref().replace('.', "_");
-        self.output.push_str("\n# TYPE ");
-        self.output.push_str(label.as_str());
-        self.output.push_str(" summary\n");
-
-        for quantile in &self.quantiles {
-            let value = h.value_at_quantile(quantile.value());
-            self.output.push_str(label.as_str());
-            self.output.push_str("{quantile=\"");
-            self.output.push_str(quantile.value().to_string().as_str());
-            self.output.push_str("\"} ");
-            self.output.push_str(value.to_string().as_str());
-            self.output.push_str("\n");
+            *sum += *value;
         }
-        self.output.push_str(label.as_str());
-        self.output.push_str("_sum ");
-        self.output.push_str(sum.to_string().as_str());
-        self.output.push_str("\n");
-        self.output.push_str(label.as_str());
-        self.output.push_str("_count ");
-        self.output.push_str(values.len().to_string().as_str());
-        self.output.push_str("\n");
     }
 }
 
 impl Clone for PrometheusRecorder {
     fn clone(&self) -> Self {
         Self {
             output: get_prom_expo_header(),
+            histos: HashMap::new(),
             quantiles: self.quantiles.clone(),
         }
     }
 }
 
 impl Into<String> for PrometheusRecorder {
```

Comment:
Same comment as before `From<PrometheusRecorder>` would be better.

Reply:
Will do.